driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -271,12 +275,7 @@ public void returnConnection(PooledConnection connection) {
             return;
         }
 
-        int inFlight = connection.inFlight.decrementAndGet();
-
-        if (trash.contains(connection)) {
-            if (inFlight == 0 && trash.remove(connection))
-                close(connection);
-        } else {
+        if (!trash.contains(connection)) {
```

Comment:
Wouldn't it be better here to just look at the connection.state instead of looking in pool.trash, or are you worried about trash and connection.state not being in sync?  I see there could be small windows where a connection could be in 'TRASH' state but not in pool.trash yet.  It looks like in both cases the connection.state will be set to TRASH before it is added to pool.trash.